From 5218d696b8df4ba4392eb6bc164ab9aa67173af8 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Fri, 04 Sep 2020 12:31:40 +0300
Subject: [PATCH] [SW-12754] gaudi: security: add partial pcie

Implement PCIe protection with "protection bits".
The RRs protection will be added in a separate commits - thus this is a
partial protection of the PCIe domain.

Change-Id: Ie72b93dc8f2f8f750f08c919e4fa37ec4674b22b
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index 25b0efd..e3c29d2 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -17,6 +17,7 @@
 
 #include <cpu_if_masks.h>
 #include <psoc_global_conf_masks.h>
+#include <pcie_wrap_masks.h>
 
 #include <preboot.h>
 #include <uart.h>
@@ -386,6 +387,52 @@
 	/* TODO: handle PSOC_DBG/PSOC_ETR/AXICTL */
 }
 
+static void preboot_secure_pci_wrap(void)
+{
+	u32 mask, prot_base, base = mmPCIE_WRAP_BASE & ~HL_CFG_BASE;
+
+	/* Set PCI level to secured */
+	mask = PCIE_WRAP_LBW_PROT_OVR_WR_EN_MASK |
+		PCIE_WRAP_LBW_PROT_OVR_RD_EN_MASK;
+	regmap_writel(mask, mmPCIE_WRAP_LBW_PROT_OVR);
+
+	prot_base = base;	/* offsets: 0 - 0x200 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 4);
+
+	prot_base += SNGL_REG_PROT_SIZE * 4;	/* offsets: 0x200 - 0x280 */
+	mask = ~(PROT_BIT(mmPCIE_WRAP_RR_ELBI_WR_SEC_REG_CTRL) &
+		PROT_BIT(mmPCIE_WRAP_RR_ELBI_RD_SEC_REG_CTRL));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	prot_base += SNGL_REG_PROT_SIZE;	/* offsets: 0x280 - 0xB00 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 17);
+
+	prot_base += SNGL_REG_PROT_SIZE * 17;	/* offsets: 0xB00 - 0xB80 */
+	mask = ~PROT_BIT(mmPCIE_WRAP_MAX_OUTSTAND);
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	prot_base += SNGL_REG_PROT_SIZE;	/* offsets: 0xB80 - 0xD00 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 3);
+
+	prot_base += SNGL_REG_PROT_SIZE;	/* offsets: 0xD00 - 0xD80 */
+	mask = ~(PROT_BIT(mmPCIE_WRAP_HBW_DRAIN_TIMEOUT) &
+		PROT_BIT(mmPCIE_WRAP_HBW_DRAIN_CFG) &
+		PROT_BIT(mmPCIE_WRAP_LBW_DRAIN_TIMEOUT) &
+		PROT_BIT(mmPCIE_WRAP_LBW_DRAIN_CFG));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	prot_base += SNGL_REG_PROT_SIZE;	/* offsets: 0xD80 - 0xF80 */
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
+static void preboot_secure_pci(void)
+{
+	/* Protect PCIE_WRAP - 4K */
+	preboot_secure_pci_wrap();
+
+	/* TODO: use RRs to protect: DBI, CORE, AUX, PHY */
+}
+
 static void preboot_security_pci_disable(void)
 {
 	regmap_writel(0, mmPCIE_WRAP_RR_WR_PRV_RANGE_REG_MIN_0);
@@ -433,6 +480,7 @@
 
 	preboot_secure_cpu();
 	preboot_secure_psoc();
+	preboot_secure_pci();
 
 	/* TODO: implement the rest of security enablement */
 
